### ISSUE TITLE
Worker is killed by SIGKILL after provider started

### ISSUE THREAD

**Issue by risinek** [openstf/stf#1211]
**What is the issue or idea you have?**
Whenever I restart provider services, it will start loading all 18 devices connected to the machine. Workers for 16 devices will start as expected, but there are always 2 devices, that won't start. Every time I restart the provider, its two random devices, not the same everytime. 
**Does it only happen on a specific device? Please run `adb devices -l` and paste the corresponding row.**
No, it happens randomly
**Please provide the steps to reproduce the issue.**
1. Have around 18 devices connected to one PC (to 5 separate USB ports via hubs (power supply is 90W - 10W max per port, and only around 4 devices connected)
2. Restart the worker
3. Each time random devices won't start and the log below will appear in provider's log stream.

**What is the expected behavior?**
Worker should not be killed and device should shows up in device list
**Do you see errors or warnings in the `stf local` output? If so, please paste them or the full log here.**
2020-03-25T18:23:00.761Z INF/provider 11 [*] Providing 0 of 18 device(s); waiting for <list of all connected device's serial#s>
...
2020-03-25T18:23:40.671Z WRN/provider 11 [*] Device worker "3a894324" was killed with signal SIGKILL, assuming deliberate action and not restarting
2020-03-25T18:23:40.671Z INF/provider 11 [*] Cleaning up device worker "3a894324"
2020-03-25T18:23:40.671Z INF/provider 11 [*] Device worker "3a894324" has retired

**Comment by denis99999** [comment]
@risinek ,
Are you sure you do not reach the USB channels capacity of your PC motherboard ?
Did you check that at USB level the 18 devices are well seen (i.e using lsusb or adb) ?

**Comment by risinek** [comment]
Yep, I can see them connected (using lsusb - already checked that). If it would be a USB capacity of motherboard, it would be always the same number of disconnected devices, won't be? Sometimes all 18 devices will get connected (rarely), next time 2-3 device workers will get killed by SIGKILL right away. 
Also every time I restart the worker, I see all 18 devices in "preparing" state in STF - that means the STF is able to communicate with all devices. But the question is, why some workers get killed by SIGKILL.
I'm gonna try running second provider on RPi3, but didn't manage to deploy it yet, so I'm looking for the cause of this problem in the meantime.

**Comment by koral--** [comment]
@risinek are there other services than STF on machine running provider?
Maybe worker processes are killed due to reaching OS process number limits?

**Comment by risinek** [comment]
Nope, there is only the provider service running in docker. I will recheck the setup what the limits are, but I doubt it would be process number limit. 
I'm also trying to figure out how to restart worker if it gets killed like this. I understand it's like this on purpose, otherwise the worker would be restarting all the time. But some manual way to restart worker for specific device serial number would do the job here.

**Comment by denis99999** [comment]
Perhaps it sounds like a hardware unstability issue due to the coupling of the hubs and the motherboard which cause the killing of the corresponding  workers.

Can you provide more information: 
- PC model and OS
- USB hub model
- full logs

Using lsusb are the 18 devices always seen even while workers are killed?

For information RPI3 is limited to 8 USB channels and one is reserved for network interface.

With RPI4 the limit is 32, I tested it connecting 23 devices through 4 USB hubs cascade-connected to only one USB 2.0 port of the RPI4 and it works fine, but without hubs cascading, the USB connections was unstable, so you can try to modify your hub topology, I recommend you one frontend hub connected to the PC and 4 backend hubs connected to the frontend one.

Also, many hub models are unstable, the best I found which is recommended by STF is the Plugable 7 ports USB 2.0 one

**Comment by risinek** [comment]
Thats a really valuable info about RPi3 and RPi4 performance, thanks! I will try to get the details about the PC. Here are hub specs - https://i-tec.cz/en/produkt/u3chargehub16-2/

**Comment by denis99999** [comment]
Look at STF readme at sections faq, troubleshooting and recommended hardware, it seems your issue is described for a part, and USB 3.0 hubs are generally not recommended due to unstability issues.

**Comment by risinek** [comment]
Ok, I'm gonna try the RPi4 as second provider. 
@denis99999 can you please provide me details about the exact model of RPi4 you have before I buy one?
Btw, the provider machine we are currently using is https://www.lenovo.com/us/en/commercial-desktop/thinkcentre/thinkcentre-m-tiny-series/M600-Tiny/p/11TC1MTM600

**Comment by denis99999** [comment]
@risinek rpi4 model B 4GB.
